src/boot/memory.c

```diff
@@ -115,139 +110,147 @@ void move_segment_table_to_dmem(void) {
 }
 #endif
 
+extern u8 __mainPoolStart[];
+
 /**
- * Initialize the main memory pool. This pool is conceptually a pair of stacks
+ * Initialize the main memory pool. This pool is conceptually regions
  * that grow inward from the left and right. It therefore only supports
  * freeing the object that was most recently allocated from a side.
  */
-void main_pool_init(void *start, void *end) {
-    sPoolStart = (u8 *) ALIGN16((uintptr_t) start) + 16;
-    sPoolEnd = (u8 *) ALIGN16((uintptr_t) end - 15) - 16;
-    sPoolFreeSpace = sPoolEnd - sPoolStart;
-
-    sPoolListHeadL = (struct MainPoolBlock *) (sPoolStart - 16);
-    sPoolListHeadR = (struct MainPoolBlock *) sPoolEnd;
-    sPoolListHeadL->prev = NULL;
-    sPoolListHeadL->next = NULL;
-    sPoolListHeadR->prev = NULL;
-    sPoolListHeadR->next = NULL;
+void main_pool_init() {
+#define SET_REGION(id, bufStart, bufEnd) \
+    gMainPool.regions[id].start = (u8 *) ALIGN4((uintptr_t)(bufStart)); \
+    gMainPool.regions[id].end = (u8 *) DOWN4((uintptr_t)(bufEnd));
+
+    SET_REGION(0, __mainPoolStart, RAM_END);
+
+#undef SET_REGION
+
 #ifdef PUPPYPRINT_DEBUG
-    mempool = sPoolFreeSpace;
+    mempool = main_pool_available();
 #endif
 }
 
-/**
- * Allocate a block of memory from the pool of given size, and from the
- * specified side of the pool (MEMORY_POOL_LEFT or MEMORY_POOL_RIGHT).
- * If there is not enough space, return NULL.
- */
-void *main_pool_alloc(u32 size, u32 side) {
-    struct MainPoolBlock *newListHead;
-    void *addr = NULL;
+// all 'try_alloc' functions expect size to be ALIGN4
 
-    size = ALIGN16(size) + 16;
-    if (size != 0 && sPoolFreeSpace >= size) {
-        sPoolFreeSpace -= size;
-        if (side == MEMORY_POOL_LEFT) {
-            newListHead = (struct MainPoolBlock *) ((u8 *) sPoolListHeadL + size);
-            sPoolListHeadL->next = newListHead;
-            newListHead->prev = sPoolListHeadL;
-            newListHead->next = NULL;
-            addr = (u8 *) sPoolListHeadL + 16;
-            sPoolListHeadL = newListHead;
-        } else {
-            newListHead = (struct MainPoolBlock *) ((u8 *) sPoolListHeadR - size);
-            sPoolListHeadR->prev = newListHead;
-            newListHead->next = sPoolListHeadR;
-            newListHead->prev = NULL;
-            sPoolListHeadR = newListHead;
-            addr = (u8 *) sPoolListHeadR + 16;
-        }
-    }
-    return addr;
+// takes the at least first 'size' bytes from 'region' and return pointer aligned on 'alignment'
+static void* main_pool_region_try_alloc_from_start_aligned(struct MainPoolRegion* region, u32 size, u32 alignment) {
+    u8* ret = (u8*) ALIGN(region->start, alignment);
+    u8* newStart = ret + size;
+    u8* regionEnd = region->end;
+    if (newStart > regionEnd)
+        return NULL;
```

Comment:
Every single one of these should have a non-debug assertion (unless any of these get significant overhead, in which debug is fine I guess).

These are usually critical enough that a failed allocation will often result in a crash, so an out of memory message will at least be more useful to see if a player runs into it and reports it.

Reply:
Added fallbacks to main pool region 0 on alloc failures + debug assert in case of failures to alloc from region 0.